### 1.3

Instead of doing analytic analysis, we use the following program to estimate the average error.

We made the following assumptions about the problem:

1. The x components follows the uniform distrubution ranging from 0 to 1.
2. Gamma is the parameter for log sum estimation, tested among [1, 0.5, 0.3, 0.2, 0.1]
3. Length is the dimension of X vectors (number of pins), tested among range(2, 10)
4. The error is defined as (estimated_value - real_max_value) / real_value
5. N = 300 is the number of samples we generate for each fixed (gamma, length) pair

In [1]:
import numpy as np

def log_sum_estimation(values, gamma=1):
    return gamma * np.log(np.sum(np.exp(values / gamma)))

np.random.seed(42)  # Setting a seed for reproducibility
N = 300 # Number of samples

for gamma in [1, 0.5, 0.3, 0.2, 0.1]:
    for length in range(2, 10):
        average_error = []
        for _ in range(N):
            sample = np.random.uniform(0, 1, length)
            estimated_value = log_sum_estimation(sample, gamma)
            real_value = np.max(sample)
            average_error.append((estimated_value - real_value) / real_value)
        average_error = sum(average_error) / len(average_error)
        print(f"Gamma: {gamma}, Length: {length}, Average Error: {average_error:.4f}")

Gamma: 1, Length: 2, Average Error: 1.2143
Gamma: 1, Length: 3, Average Error: 1.3092
Gamma: 1, Length: 4, Average Error: 1.5269
Gamma: 1, Length: 5, Average Error: 1.6541
Gamma: 1, Length: 6, Average Error: 1.7562
Gamma: 1, Length: 7, Average Error: 1.8681
Gamma: 1, Length: 8, Average Error: 2.0005
Gamma: 1, Length: 9, Average Error: 2.0315
Gamma: 0.5, Length: 2, Average Error: 0.5222
Gamma: 0.5, Length: 3, Average Error: 0.5439
Gamma: 0.5, Length: 4, Average Error: 0.6435
Gamma: 0.5, Length: 5, Average Error: 0.6867
Gamma: 0.5, Length: 6, Average Error: 0.7310
Gamma: 0.5, Length: 7, Average Error: 0.7804
Gamma: 0.5, Length: 8, Average Error: 0.8279
Gamma: 0.5, Length: 9, Average Error: 0.8704
Gamma: 0.3, Length: 2, Average Error: 0.2516
Gamma: 0.3, Length: 3, Average Error: 0.2718
Gamma: 0.3, Length: 4, Average Error: 0.3010
Gamma: 0.3, Length: 5, Average Error: 0.3232
Gamma: 0.3, Length: 6, Average Error: 0.3412
Gamma: 0.3, Length: 7, Average Error: 0.3814
Gamma: 0.3, Length: 8, Ave

The result is as follows:

| Gamma | Length: 2 | Length: 3 | Length: 4 | Length: 5 | Length: 6 | Length: 7 | Length: 8 | Length: 9 |
|-------|-----------|-----------|-----------|-----------|-----------|-----------|-----------|-----------|
| 1     | 1.2143    | 1.3092    | 1.5269    | 1.6541    | 1.7562    | 1.8681    | 2.0005    | 2.0315    |
| 0.5   | 0.5222    | 0.5439    | 0.6435    | 0.6867    | 0.7310    | 0.7804    | 0.8279    | 0.8704    |
| 0.3   | 0.2516    | 0.2718    | 0.3010    | 0.3232    | 0.3412    | 0.3814    | 0.3946    | 0.4254    |
| 0.2   | 0.1078    | 0.1447    | 0.1380    | 0.1730    | 0.1757    | 0.1948    | 0.2156    | 0.2291    |
| 0.1   | 0.0530    | 0.0387    | 0.0409    | 0.0457    | 0.0527    | 0.0542    | 0.0633    | 0.0701    |

The smaller the value of gamma, the smaller of the input lengh, the less the average error. Also, as gamma approaches zero, the average error also approaches zero.